play.hs

```diff
@@ -0,0 +1,154 @@
+-- CPSC 312 - 2019 - Project 1 - Sokoban
+-- Authors: Rebecca Li, Jason Doo, Gurveer Aulakh
+
+-- To play:
+-- ghci
+-- :l play
+-- go
+
+module Play where
+import System.IO
+import Data.Char
+import Sokoban
+import Hints
+
+-- [TODO]: ?
+-- passing the state instead of board and (pr,pc) separately to move & push functions?
+
+{-- Game Functions --}
+
+-- main play function
+play :: Result -> [Coordinates] -> IO()
+play (ContinueGame s) deadsquares =
+  do
+    printBoard (board s)
+    putStr "Input your next move (W,A,S,D):"
+    move <- getLine 
+    case move of
+      [] -> play (ContinueGame s) deadsquares -- continue if no input
+      _ -> play (getNextBoard (Action (toUpper (move!!0))) s deadsquares) deadsquares      
+
+play (WonGame s) _ =
+  do
+    printBoard (board s)
+    putStrLn "Congratulations! You beat the level!"
+
+play (QuitGame s) _ =
+  do
+    putStrLn "Exiting level..."
+
+play (UndoGame (State board pS (r, c) l)) deadsquares =
+  do
+    putStrLn "Switching to previous move..."
+    if pS == Empty then (putStrLn "No previous move...")
+    else play (ContinueGame pS) deadsquares
+     
+play (RestartGame (State board pS (r, c) l)) _ =
+  do
+    putStrLn "Restarting level..."
+    playLevel l
+
+-- takes user input and current state of board
+-- returns the next state of board as a result (game won or continue game)
+-- should be used as a helper in the main game function
+getNextBoard :: Action -> State -> [Coordinates] -> Result
+getNextBoard move (State board pS (r,c) l) deadsquares
+  | move == (Action 'W') = movePlayer (State board newPrevState (r,c) l) (r-1,c) (r-2,c)
+  | move == (Action 'A') = movePlayer (State board newPrevState (r,c) l) (r,c-1) (r,c-2)
+  | move == (Action 'S') = movePlayer (State board newPrevState (r,c) l) (r+1,c) (r+2,c)
+  | move == (Action 'D') = movePlayer (State board newPrevState (r,c) l) (r,c+1) (r,c+2)
+  | move == (Action 'H') = if (isGameWon (giveHint (State board pS (r,c) l) deadsquares))
+                               then WonGame (giveHint (State board pS (r,c) l) deadsquares)
+                               else ContinueGame (giveHint (State board pS (r,c) l) deadsquares)
```

Comment:
having to pass around `deadsquares` everywhere all the time, despite only needing it when we do the hint seems a bit awkward now. I see that this way we can compute it just once, but maybe we can just have a separate helper that returns dead squares given a level (computing dead squares seem efficient). Or we could pre-compute and store it like:
```
getDeadSquares level  
  | level == 1 = [(1,1)]
  | level == 2 = [(1,1),(1,2),(1,3),(1,4),(2,1),(4,2),(5,5),(6,1),(6,6),(7,1),(7,6),(8,5),(8,6)]
  | level == 3 = [(1,1),(1,2),(1,3),(1,4),(1,5),(1,6),(2,1),(3,1),(4,1),(5,1),(5,2),(5,3),(5,4),(5,5),(5,6)]
  | ...
```
since we will always have a set amount of levels for the game. What do you think?

Reply:
Yeah, that works. I was thinking about the TA saying not to hardcode the solutions--I don't know if that applies to this?

Reply:
It doesn't. giveHint now calls the findDeadSquares function (deadsquares no longer passed in play.hs)